wire-schema/src/main/java/com/squareup/wire/schema/Field.kt

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+import com.google.common.collect.ImmutableList
+import com.squareup.wire.schema.Options.FIELD_OPTIONS
+import com.squareup.wire.schema.internal.parser.FieldElement
+
+class Field private constructor(
+  val packageName: String?,
+
+  val location: Location,
+
+  /** May be null for proto3 fields. */
+  val label: Label?,
+
+  val name: String,
+
+  val documentation: String,
+
+  val tag: Int,
+
+  val default: String?,
+
+  private val elementType: String,
+
+  val options: Options,
+
+  val isExtension: Boolean
+) {
+  // Null until this field is linked.
+  var type: ProtoType? = null
+    private set
+
+  // Null until this field is linked.
+  private var deprecated: Any? = null
+
+  // Null until this field is linked.
+  private var packed: Any? = null
+
+  // Null until this field is linked.
+  var isRedacted: Boolean = false
+    private set
+
+  val isRepeated: Boolean
+    get() = label == Label.REPEATED
+
+  val isOptional: Boolean
+    get() = label == Label.OPTIONAL
+
+  val isRequired: Boolean
+    get() = label == Label.REQUIRED
+
+  /**
+   * Returns this field's name, prefixed with its package name. Uniquely identifies extension
+   * fields, such as in options.
+   */
+  val qualifiedName: String
+    get() {
+      return when {
```

Comment:
Compute once and assign? Same for fields below.

Reply:
I think it reads better as computed-on-demand state. Lets discuss offline; I’m don’t hold strong feelings here.